src/v/utils/external_process.cc

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "external_process.h"
+
+#include <seastar/core/loop.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/timed_out_error.hh>
+#include <seastar/core/when_all.hh>
+#include <seastar/core/with_timeout.hh>
+#include <seastar/util/later.hh>
+#include <seastar/util/process.hh>
+
+#include <functional>
+#include <optional>
+
+namespace external_process {
+ss::future<std::unique_ptr<external_process>>
+external_process::create_external_process(
+  std::vector<ss::sstring> command,
+  std::optional<std::vector<ss::sstring>> env) {
+    vassert(
+      !command.empty(),
+      "Should not be passed an empty command list to "
+      "create_external_process");
+
+    auto path = std::filesystem::path(command[0]);
+    ss::experimental::spawn_parameters params{
+      .argv = std::move(command),
+      .env = env.has_value() ? std::move(env).value()
+                             : std::vector<ss::sstring>{}};
+
+    auto p = co_await ss::experimental::spawn_process(path, params);
+
+    co_return std::unique_ptr<external_process>(
+      new external_process(std::move(p), std::move(path), std::move(params)));
+}
+
+external_process::external_process(
+  ss::experimental::process p,
+  std::filesystem::path path,
+  ss::experimental::spawn_parameters params)
+  : _process(std::move(p))
+  , _path(std::move(path))
+  , _params(std::move(params))
+  , _terminate_mutex("external_process::terminate")
+  , _wait_mutex("external_process::wait") {}
+
+external_process::~external_process() {
+    vassert(_completed, "Process has not yet completed");
+    vassert(!_stdout_consumer.has_value(), "stdout consumer not cleaned up");
+    vassert(!_stderr_consumer.has_value(), "stderr consumer not cleaned up");
+}
+
+ss::future<ss::experimental::process::wait_status> external_process::wait() {
+    auto guard = _gate.hold();
+    if (_completed) {
+        throw std::system_error(
+          make_error_code(error_code::process_already_completed));
```

Comment:
nitpick: `make_error_code` isn't needed (there are several of these)
```suggestion
        throw std::system_error(error_code::process_already_completed);
```

Reply:
`make_error_code` will return a std::error_code that contains the integer code and the system_category, which I think we want included.  `error_code` here is just an enum.

Reply:
oh duh right the `is_error_code_enum` struct does the implicit conversion